src/common/histogram.c

```diff
@@ -29,332 +29,163 @@
+                                  const dt_iop_order_iccprofile_info_t *const profile));
 
-//------------------------------------------------------------------------------
-
-inline static void histogram_helper_cs_RAW_helper_process_pixel_float(
-    const dt_dev_histogram_collection_params_t *const histogram_params, const float *pixel, uint32_t *histogram)
+static inline void _clamp_bin(const dt_aligned_pixel_t vals,
+                              uint32_t *const restrict histogram,
+                              const float max_bin)
 {
-  const uint32_t i = PS(*pixel, histogram_params);
-  histogram[4 * i]++;
-}
-
-inline static void histogram_helper_cs_RAW(const dt_dev_histogram_collection_params_t *const histogram_params,
-                                           const void *pixel, uint32_t *histogram, int j,
-                                           const dt_iop_order_iccprofile_info_t *const profile_info)
-{
-  const dt_histogram_roi_t *roi = histogram_params->roi;
-  const float *input = (float *)pixel + roi->width * j + roi->crop_x;
-  for(int i = 0; i < roi->width - roi->crop_width - roi->crop_x; i++, input++)
-  {
-    histogram_helper_cs_RAW_helper_process_pixel_float(histogram_params, input, histogram);
-  }
+  DT_ALIGNED_PIXEL size_t bin[4];
+  for_each_channel(k,aligned(vals,bin:16))
+    // must be signed before clamping as value may be negative
+    bin[k] = CLAMP(vals[k], 0.0f, max_bin);
+
+  histogram[bin[0]*4]++;
+  histogram[bin[1]*4+1]++;
+  histogram[bin[2]*4+2]++;
 }
 
 //------------------------------------------------------------------------------
 
-// WARNING: you must ensure that bins_count is big enough
-inline static void histogram_helper_cs_RAW_helper_process_pixel_uint16(
-    const dt_dev_histogram_collection_params_t *const histogram_params, const uint16_t *pixel, uint32_t *histogram)
+static inline void _bin_raw(const dt_dev_histogram_collection_params_t *const params,
+                            const void *pixel, uint32_t *histogram, int j,
+                            const dt_iop_order_iccprofile_info_t *const profile)
 {
-  const uint16_t i = PU(*pixel, histogram_params);
-  histogram[4 * i]++;
-}
-
-inline void dt_histogram_helper_cs_RAW_uint16(const dt_dev_histogram_collection_params_t *const histogram_params,
-                                              const void *pixel, uint32_t *histogram, int j,
-                                              const dt_iop_order_iccprofile_info_t *const profile_info)
-{
-  const dt_histogram_roi_t *roi = histogram_params->roi;
+  const dt_histogram_roi_t *roi = params->roi;
   uint16_t *in = (uint16_t *)pixel + roi->width * j + roi->crop_x;
+  const size_t max_bin = params->bins_count - 1;
 
-  // process pixels
-  for(int i = 0; i < roi->width - roi->crop_width - roi->crop_x; i++, in++)
-    histogram_helper_cs_RAW_helper_process_pixel_uint16(histogram_params, in, histogram);
+  for(int i = 0; i < roi->width - roi->crop_width - roi->crop_x; i++)
```

Comment:
please enclose with `{ }` the loop instructions. With the comment in between the first line of code it is a source of errors for maintenance.